src/Psy/Shell.php

```diff
@@ -74,6 +76,11 @@ public function __construct(Configuration $config = null)
 
         parent::__construct('Psy Shell', self::VERSION);
 
+        if ($this->config->getTabCompletion()) {
+            $this->completion = new AutoCompleter($this->context, $this->all());
+            $this->completion->activate();
+        }
```

Comment:
I think I'd prefer to move instantiation of this service to Configuration, where other services are instantiated. That way users can interact with it in the psysh config file, much the same way they can interact with commands, presenters and other extension points.

Then I could return this from my config:

```php
return [
  'tabCompletionMatchers' => [
    new EloquentModelMatcher(),
    new FancyPantsMatcher()
  ]
]
```

… and get some custom matchers that meet my needs, to go with the default set.

Along with that we would probably have to change to a ContextAwareInterface style dependency injection for the matchers that depend on context, shell, etc, so that users wouldn't have to have access to those things when setting up the initial configuration. Commands already do this (see ReflectingCommand, WtfCommand).

Maybe we use the config as an autocompleter factory, and just activate it here?

```php
if ($this->config->getTabCompletion()) {
  $this->completion = $this->config->getAutoCompleter();
  $this->completion->activate();
}
```

I haven't thought through this completely, but it seems like a direction we should head :)